Drv/Ip/IpSocket.cpp

```diff
@@ -0,0 +1,204 @@
+    this->m_timeoutSeconds = timeout_seconds;
+    this->m_timeoutMicroseconds = timeout_microseconds;
+    this->m_port = port;
+    (void) Fw::StringUtils::string_copy(this->m_hostname, hostname, SOCKET_MAX_HOSTNAME_SIZE);
+    return SOCK_SUCCESS;
+}
+
+SocketIpStatus IpSocket::setupTimeouts(NATIVE_INT_TYPE socketFd) {
+// Get the IP address from host
+#ifdef TGT_OS_TYPE_VXWORKS
+    // No timeouts set on Vxworks
+#else
+    // Set timeout socket option
+    struct timeval timeout;
+    timeout.tv_sec = this->m_timeoutSeconds;
+    timeout.tv_usec = this->m_timeoutMicroseconds;
+    // set socket write to timeout after 1 sec
+    if (setsockopt(socketFd, SOL_SOCKET, SO_SNDTIMEO, (char*)&timeout, sizeof(timeout)) < 0) {
+        return SOCK_FAILED_TO_SET_SOCKET_OPTIONS;
+    }
+    return SOCK_SUCCESS;
+#endif
+}
+
+SocketIpStatus IpSocket::addressToIp4(const char* address, void* ip4) {
+    FW_ASSERT(address != NULL);
+    FW_ASSERT(ip4 != NULL);
+    // Get the IP address from host
+#ifdef TGT_OS_TYPE_VXWORKS
+    U32 ip = inet_addr(address);
+    if (ip == ERROR) {
+        return SOCK_INVALID_IP_ADDRESS;
+    }
+    *ip4 = ip;
+#else
+    // First IP address to socket sin_addr
+    if (not ::inet_pton(AF_INET, address, ip4)) {
+        return SOCK_INVALID_IP_ADDRESS;
+    };
+#endif
+    return SOCK_SUCCESS;
+}
+
+bool IpSocket::isOpened(void) {
+    bool is_open = false;
+    m_lock.lock();
+    is_open = m_open;
+    m_lock.unLock();
+    return is_open;
+}
+
+void IpSocket::close(void) {
+    m_lock.lock();
+    if (this->m_fd != -1) {
+        (void)::shutdown(this->m_fd, SHUT_RDWR);
+        (void)::close(this->m_fd);
+        this->m_fd = -1;
+    }
+    m_open = false;
+    m_lock.unLock();
+}
+
+SocketIpStatus IpSocket::open(void) {
+    NATIVE_INT_TYPE fd = -1;
+    SocketIpStatus status = SOCK_SUCCESS;
+    FW_ASSERT(m_fd == -1 and not m_open); // Ensure we are not opening an opened socket
+    // Open a TCP socket for incoming commands, and outgoing data if not using UDP
+    status = this->openProtocol(fd);
+    if (status != SOCK_SUCCESS) {
+        FW_ASSERT(m_fd == -1); // Ensure we properly kept closed on error
+        return status;
+    }
+    // Lock to update values and "officially open"
+    m_lock.lock();
+    m_fd = fd;
+    m_open = true;
+    m_lock.unLock();
+    return status;
+}
+
+SocketIpStatus IpSocket::send(const U8* const data, const U32 size) {
+    U32 total = 0;
+    I32 sent  = 0;
+    // Prevent transmission before connection, or after a disconnect
+    if (this->m_fd == -1) {
+        return SOCK_DISCONNECTED;
+    }
+    // Attempt to send out data and retry as necessary
+    for (U32 i = 0; (i < SOCKET_MAX_ITERATIONS) && (total < size); i++) {
+        sent = 0;
+        // Send using my specific protocol
+        sent = this->sendProtocol(data + total, size - total);
+        // Error is EINTR or timeout just try again
+        if (((sent == -1) && (errno == EINTR)) || (sent == 0)) {
```

Comment:
Don't think this is possible.  If the `send` variable is zero, it will skip the loop entirely and won't check that condition.  Thus if a user sends 0 bytes, no send call will be triggered.

Reply:
I don't understand your comment about the timeout.  In the case of a partial send, we just retry the next bytes.  In the case of a timeout on the zeroth byte we retry again.  This will work until we hit the maximum retries....and then we report to the user.

Reply:
> Don't think this is possible. If the send variable is zero, it will skip the loop entirely and won't check that condition. Thus if a user sends 0 bytes, no send call will be triggered.

Looking at the loop condition I agree. My concern was getting stuck in the loop if the user provided a send size of zero, but since we never meet the conditions to enter the loop if the size is 0 that isn't a concern.

> I don't understand your comment about the timeout. In the case of a partial send, we just retry the next bytes. In the case of a timeout on the zeroth byte we retry again. This will work until we hit the maximum retries....and then we report to the user.

Ok, with your understanding I now agree the code is correctly handling the timeout case.

Reply:
Thinking about it more however, do we really want to transparently handle timeouts? If a user is applying a timeout, my suspicion is that they could be concerned about meeting timing deadlines. If I set a send timeout, I would expect the send call to return within the timeout window to make sure I don't miss other deadlines.

Users could resolve this by dropping down the number of retries, but tweaking this number also affects the number of EINTR signal interrupts that could be received by both send and receive.

Also the defaults as currently set makes it really easy to shoot yourself in the foot. With `0xFF` retries you effectively have a near infinite timeout.

I think I understand your goal of trying to transparently handle timeouts for users, which is useful, I'm just not sure that's what all users setting a timeout want. Some users may want to re-try a timeout, others may want to just treat it as an error and move on.

Reply:
If the user wants the call to timeout and return within a specific interval, they may simply set the MAX_ITERATIONS configuration to 1 and it will return a `SOCK_INTERRUPTED_TRY_AGAIN`.   

The default paradigm is meant to represent the case of "`send` will send all data and will block until doing so" without the risk of a literal block forever scenario.  This parallels the `recv` functionality which **does** block forever, but is protected on its own thread.  `send` does not have a thread and thus should have a finite limit.

Reply:
In re-reviewing this, I think I see your point, which is that interrupts and send timeouts are coupled together when they likely should not be.  Send timeouts and retries in the case of interruptions should be handled differently.  I'll see if there is a clean way to pull these apart.

Reply:
I'm seeing a few quirks with the current timeout design and I just want to make sure we're on the same page.
Do you think are worth resolving?

- EINTR and TIMEOUT retries are tied together, so if you're using signals you can't safely drop the timeout retries to zero.
- SOCK_INTERRUPTED_TRY_AGAIN has no way of returning how much data was actually sent, so that makes it difficult to handle partial sends after a timeout.
- Users that care about timeouts need to know to set both the timeout and the retry count, otherwise the high retry count will cause send to still effectively block. This seems easy to miss and could be painful to debug.

Reply:
This code supports specifically the F´components that wrap these items.  Thus we don't have to support "whatever users might want" but more narrowly "what our supplied components expect".    Given this and the fact that the tcp driver is inline with other drivers I'd recommend leaving the code as is.

Some notes:
1. `send timeouts` are reset if a byte is sent, thus the "total time" is inherently hard to predict.
2. Retries are the standard mechanism for dealing with a message that does not send, so iterating a large fixed time is expected.
3. Other drivers in the driver package exhibit this same block until done behavior based on the assumption that `send` will usually succeed.

I'd recommend breaking out clean driver timeouts into its own issue that we can revisit at a later time as it will involve refactoring all of our Linux driver code.  There is a way to implement this correctly, but it is more complicated than should be included here.

Reply:
> This code supports specifically the F´components that wrap these items. Thus we don't have to support "whatever users might want" but more narrowly "what our supplied components expect".

Yes, but if don't expose this functionality to the TCP and UDP components users don't have access to this functionality either. 😄 

As you've said, cleanly handling timeouts isn't an easy problem to solve. If `SO_SNDTIMEO` doesn't provide a guaranteed upper time bound it doesn't seem like there's a lot of benefit for F' for properly handling them either.

I'm happy to punt this problem for later if you're okay with it, I was just making a big deal out of it now since it seems like it might require interface changes which will be more difficult to make later on.

Reply:
The drivers have an enumeration that lists error codes. It would be fairly straight forward to update that to add a TIMEOUT error code without the changing the API.  Thus, I am not concerned with splitting this work out into a new issue for further down the road.